File: modules/crypto-core/src/main/java/com/shterneregen/securelan/crypto/keystore/impl/DefaultKeyStoreService.java
package com.shterneregen.securelan.crypto.keystore.impl;

import com.shterneregen.securelan.crypto.keystore.KeyStoreService;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.cert.Certificate;
import java.util.Objects;

public final class DefaultKeyStoreService implements KeyStoreService {
    private static final String TYPE = "PKCS12";

    @Override
    public KeyStore createPkcs12() {
        try {
            KeyStore keyStore = KeyStore.getInstance(TYPE);
            keyStore.load(null, null);
            return keyStore;
        } catch (IOException | GeneralSecurityException exception) {
            throw new IllegalStateException("Failed to create PKCS12 keystore", exception);
        }
    }

    @Override
    public KeyStore load(Path path, char[] password) throws IOException, GeneralSecurityException {
        Objects.requireNonNull(path, "path");
        KeyStore keyStore = KeyStore.getInstance(TYPE);
        try (InputStream inputStream = Files.newInputStream(path)) {
            keyStore.load(inputStream, password);
        }
        return keyStore;
    }

    @Override
    public void store(KeyStore keyStore, Path path, char[] password) throws IOException, GeneralSecurityException {
        Objects.requireNonNull(keyStore, "keyStore");
        Objects.requireNonNull(path, "path");
        Path parent = path.getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }
        try (OutputStream outputStream = Files.newOutputStream(path)) {
            keyStore.store(outputStream, password);
        }
    }

    @Override
    public void setPrivateKeyEntry(KeyStore keyStore,
                                   String alias,
                                   PrivateKey privateKey,
                                   char[] entryPassword,
                                   Certificate[] chain) throws GeneralSecurityException {
        Objects.requireNonNull(keyStore, "keyStore");
        keyStore.setKeyEntry(alias, privateKey, entryPassword, chain);
    }

    @Override
    public void setCertificateEntry(KeyStore keyStore, String alias, Certificate certificate) throws GeneralSecurityException {
        Objects.requireNonNull(keyStore, "keyStore");
        keyStore.setCertificateEntry(alias, certificate);
    }

    @Override
    public PrivateKey getPrivateKey(KeyStore keyStore, String alias, char[] entryPassword) throws GeneralSecurityException {
        Objects.requireNonNull(keyStore, "keyStore");
        return (PrivateKey) keyStore.getKey(alias, entryPassword);
    }

    @Override
    public PublicKey getPublicKey(KeyStore keyStore, String alias) throws GeneralSecurityException {
        Certificate certificate = getCertificate(keyStore, alias);
        return certificate == null ? null : certificate.getPublicKey();
    }

    @Override
    public Certificate getCertificate(KeyStore keyStore, String alias) throws GeneralSecurityException {
        Objects.requireNonNull(keyStore, "keyStore");
        return keyStore.getCertificate(alias);
    }
}

File: modules/crypto-core/src/main/kotlin/com/shterneregen/securelan/crypto/keystore/impl/DefaultKeyStoreService.kt
package com.shterneregen.securelan.crypto.keystore.impl

import com.shterneregen.securelan.crypto.keystore.KeyStoreService
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.security.GeneralSecurityException
import java.security.KeyStore
import java.security.PrivateKey
import java.security.PublicKey
import java.security.cert.Certificate
import java.util.Objects

class DefaultKeyStoreService : KeyStoreService {
    override fun createPkcs12(): KeyStore = try {
        KeyStore.getInstance(TYPE).apply { load(null, null) }
    } catch (exception: IOException) {
        throw IllegalStateException("Failed to create PKCS12 keystore", exception)
    } catch (exception: GeneralSecurityException) {
        throw IllegalStateException("Failed to create PKCS12 keystore", exception)
    }

    @Throws(IOException::class, GeneralSecurityException::class)
    override fun load(path: Path, password: CharArray): KeyStore {
        Objects.requireNonNull(path, "path")
        val keyStore = KeyStore.getInstance(TYPE)
        Files.newInputStream(path).use { inputStream -> keyStore.load(inputStream, password) }
        return keyStore
    }

    @Throws(IOException::class, GeneralSecurityException::class)
    override fun store(keyStore: KeyStore, path: Path, password: CharArray) {
        Objects.requireNonNull(keyStore, "keyStore")
        Objects.requireNonNull(path, "path")
        path.parent?.let { Files.createDirectories(it) }
        Files.newOutputStream(path).use { outputStream -> keyStore.store(outputStream, password) }
    }

    @Throws(GeneralSecurityException::class)
    override fun setPrivateKeyEntry(
        keyStore: KeyStore,
        alias: String,
        privateKey: PrivateKey,
        entryPassword: CharArray,
        chain: Array<Certificate>,
    ) {
        Objects.requireNonNull(keyStore, "keyStore")
        keyStore.setKeyEntry(alias, privateKey, entryPassword, chain)
    }

    @Throws(GeneralSecurityException::class)
    override fun setCertificateEntry(keyStore: KeyStore, alias: String, certificate: Certificate) {
        Objects.requireNonNull(keyStore, "keyStore")
        keyStore.setCertificateEntry(alias, certificate)
    }

    @Throws(GeneralSecurityException::class)
    override fun getPrivateKey(keyStore: KeyStore, alias: String, entryPassword: CharArray): PrivateKey {
        Objects.requireNonNull(keyStore, "keyStore")
        return keyStore.getKey(alias, entryPassword) as PrivateKey
    }

    @Throws(GeneralSecurityException::class)
    override fun getPublicKey(keyStore: KeyStore, alias: String): PublicKey? = getCertificate(keyStore, alias)?.publicKey

    @Throws(GeneralSecurityException::class)
    override fun getCertificate(keyStore: KeyStore, alias: String): Certificate? {
        Objects.requireNonNull(keyStore, "keyStore")
        return keyStore.getCertificate(alias)
    }

    companion object {
        private const val TYPE = "PKCS12"
    }
}
